File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/diskdrive/FluidDriveWrapperStorageDisk.java
package com.refinedmods.refinedstorage.apiimpl.network.node.diskdrive;

import com.refinedmods.refinedstorage.api.storage.AccessType;
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDisk;
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskContainerContext;
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskListener;
import com.refinedmods.refinedstorage.api.util.Action;
import com.refinedmods.refinedstorage.apiimpl.network.node.DiskState;
import com.refinedmods.refinedstorage.blockentity.config.IWhitelistBlacklist;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.fluids.FluidStack;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Collection;
import java.util.UUID;

public class FluidDriveWrapperStorageDisk implements IStorageDisk<FluidStack> {
    private final DiskDriveNetworkNode diskDrive;
    private final IStorageDisk<FluidStack> parent;
    private DiskState lastState;

    public FluidDriveWrapperStorageDisk(DiskDriveNetworkNode diskDrive, IStorageDisk<FluidStack> parent) {
        this.diskDrive = diskDrive;
        this.parent = parent;
        this.setSettings(
            () -> {
                DiskState currentState = DiskState.get(getStored(), getCapacity());

                if (this.lastState != currentState) {
                    this.lastState = currentState;

                    diskDrive.requestBlockUpdate();
                }
            },
            diskDrive
        );
        this.lastState = DiskState.get(getStored(), getCapacity());
    }

    @Override
    public int getPriority() {
        return diskDrive.getPriority();
    }

    @Override
    public AccessType getAccessType() {
        return parent.getAccessType();
    }

    @Override
    public Collection<FluidStack> getStacks() {
        return parent.getStacks();
    }

    @Override
    @Nonnull
    public FluidStack insert(@Nonnull FluidStack stack, int size, Action action) {
        if (!IWhitelistBlacklist.acceptsFluid(diskDrive.getFluidFilters(), diskDrive.getWhitelistBlacklistMode(), diskDrive.getCompare(), stack)) {
            return StackUtils.copy(stack, size);
        }

        return parent.insert(stack, size, action);
    }

    @Override
    @Nonnull
    public FluidStack extract(@Nonnull FluidStack stack, int size, int flags, Action action) {
        return parent.extract(stack, size, flags, action);
    }

    @Override
    public int getStored() {
        return parent.getStored();
    }

    @Override
    public int getCacheDelta(int storedPreInsertion, int size, @Nullable FluidStack remainder) {
        return parent.getCacheDelta(storedPreInsertion, size, remainder);
    }

    @Override
    public int getCapacity() {
        return parent.getCapacity();
    }

    @Nullable
    @Override
    public UUID getOwner() {
        return parent.getOwner();
    }

    @Override
    public void setSettings(@Nullable IStorageDiskListener listener, IStorageDiskContainerContext context) {
        parent.setSettings(listener, context);
    }

    @Override
    public CompoundTag writeToNbt() {
        return parent.writeToNbt();
    }

    @Override
    public ResourceLocation getFactoryId() {
        return parent.getFactoryId();
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/network/node/diskdrive/FluidDriveWrapperStorageDisk.kt
package com.refinedmods.refinedstorage.apiimpl.network.node.diskdrive

import com.refinedmods.refinedstorage.api.storage.AccessType
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDisk
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskContainerContext
import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskListener
import com.refinedmods.refinedstorage.api.util.Action
import com.refinedmods.refinedstorage.apiimpl.network.node.DiskState
import com.refinedmods.refinedstorage.tile.config.IWhitelistBlacklist
import com.refinedmods.refinedstorage.util.StackUtils.copy
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.Identifier
import net.minecraftforge.fluids.FluidInstance


class FluidDriveWrapperStorageDisk(private val diskDrive: DiskDriveNetworkNode, parent: IStorageDisk<FluidInstance>?) : IStorageDisk<FluidInstance?> {
    private val parent: IStorageDisk<FluidInstance>?
    private var lastState: DiskState
    override fun getPriority(): Int {
        return diskDrive.priority
    }

    override fun getAccessType(): AccessType? {
        return parent!!.getAccessType()
    }

    override fun getStacks(): Collection<FluidInstance>? {
        return parent!!.getStacks()
    }

    @Nonnull
    override fun insert(@Nonnull stack: FluidInstance, size: Int, action: Action?): FluidInstance {
        return if (!IWhitelistBlacklist.acceptsFluid(diskDrive.fluidFilters, diskDrive.whitelistBlacklistMode, diskDrive.compare, stack)) {
            copy(stack, size)
        } else parent!!.insert(stack, size, action)
    }

    @Nonnull
    override fun extract(@Nonnull stack: FluidInstance, size: Int, flags: Int, action: Action?): FluidInstance {
        return parent!!.extract(stack, size, flags, action)
    }

    override fun getStored(): Int {
        return parent!!.getStored()
    }

    override fun getCacheDelta(storedPreInsertion: Int, size: Int, @Nullable remainder: FluidInstance): Int {
        return parent!!.getCacheDelta(storedPreInsertion, size, remainder)
    }

    override val capacity: Int
        get() = parent!!.capacity

    override fun setSettings(@Nullable listener: IStorageDiskListener?, context: IStorageDiskContainerContext) {
        parent!!.setSettings(listener, context)
    }

    override fun writeToNbt(): CompoundTag? {
        return parent!!.writeToNbt()
    }

    override val factoryId: Identifier
        get() = parent!!.factoryId

    init {
        this.parent = parent
        setSettings(
                IStorageDiskListener {
                    val currentState: DiskState = DiskState.Companion.get(getStored(), capacity)
                    if (lastState != currentState) {
                        lastState = currentState
                        diskDrive.requestBlockUpdate()
                    }
                },
                diskDrive
        )
        lastState = DiskState.Companion.get(getStored(), capacity)
    }
}